### PR TITLE
client: get caller's uid/gid on every libcephfs operation

### PR DESCRIPTION
**Pull request by ukernel**
Fixes: http://tracker.ceph.com/issues/17591
Signed-off-by: Yan, Zheng <zyan@redhat.com>

### CODE REVIEW

**Comment by jcsp** [comment]
Looks sane to me, @gregsfortytwo please could you review

**Comment by gregsfortytwo** [comment]
I was really trying to move away from this model, but I guess until we have @jtlayton's new libcephfs and bindings done we need our tests to pass. :/ Did we run this through the samba test suite?

File: src/client/UserPerm.h [hunk 1]
```diff
@@ -65,8 +65,8 @@ struct UserPerm
     return *this;
   }
 
-  uid_t uid() const { return m_uid; }
-  gid_t gid() const { return m_gid; }
+  uid_t uid() const { return m_uid != (uid_t)-1 ? m_uid : ::geteuid(); }
+  gid_t gid() const { return m_gid != (gid_t)-1 ? m_gid : ::getegid(); }
```
Comment (jtlayton) [inline src/client/UserPerm.h]: For the record, we should really be using the fsuid and fsgid here (at least on Linux) and not the euid/egid. In most cases, they are identical but there are cases where they can be different. Unfortunately, Linux has no getfsuid() syscall -- you have to issue setfsuid(-1) which will always fail, but will return the existing fsuid.

What we should probably do is roll our own getfsuid() routine that does that on Linux and just calls out to geteuid() on other platforms. That said, we should probably do that in the context of a different PR.

Comment (jtlayton) [inline src/client/UserPerm.h]: Hmm...maybe I'm wrong there actually. The setfsuid manpage says "don't use it", so maybe geteuid is fine. I think you can disregard my comment...

**Comment by ukernel** [comment]
http://pulpito.ceph.com/zyan-2016-10-20_07:49:59-samba-wip-zyan-testing---basic-smithi/

**Pull request merged**